### PR TITLE
feat: add functions as UDF input types

### PR DESCRIPTION
**Pull request by stevenpyzhang**
### Description 
Add functions as UDF argument types. Lambdas are still not usable though since we need to complete work in ExpressionTypeManager , CodeGenRunner, and SqlToJavaVisitor. (Coming soon)

Added in TRANSFORM_ARRAY, TRANSFORM_MAP, REDUCE_MAP

### Testing done 
Unit tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Review by agavra — APPROVED** [review]
Excellent. Looking forward to the next PRs in the chain!

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/map/MapReduce.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Reduce a map using an initial state and function
+ */
+@UdfDescription(
+    name = "map_reduce",
+    category = FunctionCategory.MAP,
+    description = "Reduce the input map down to a single value " 
+        + "using an initial state and a function. " 
+        + "The initial state (s) and is passed into the scope of the function. " 
+        + "Each invocation returns a new value for s, which the next invocation will receive."
```
Comment (guozhangwang) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/map/MapReduce.java:36]: nit: space after "."

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/map/MapReduce.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * Reduce a map using an initial state and function
+ */
+@UdfDescription(
+    name = "map_reduce",
+    category = FunctionCategory.MAP,
+    description = "Reduce the input map down to a single value " 
+        + "using an initial state and a function. " 
+        + "The initial state (s) and is passed into the scope of the function. "
```
Comment (guozhangwang) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/map/MapReduce.java:35]: nit: "and"?

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/map/MapTransform.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ */
+@UdfDescription(
+    name = "map_transform",
+    category = FunctionCategory.MAP,
+    description = "Apply one function to each key and " 
+        + "one function to each value of a map. " 
+        + "The two arguments for each function are in order: key, value. " 
+        + "The first function provided will be applied to each key and the " 
+        + "second one applied to each value. "
+        + "The transformed map is returned",
```
Comment (guozhangwang) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/map/MapTransform.java:38]: nit: "is returned."

**Review by guozhangwang — APPROVED** [review]
Just nits, otherwise LGTM.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/map/MapReduce.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import io.confluent.ksql.function.udf.UdfParameter;
+import io.confluent.ksql.util.KsqlConstants;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * Reduce a map using an initial state and function
+ */
+@UdfDescription(
+    name = "map_reduce",
```
Comment (rodesai) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/map/MapReduce.java:31]: can we name this `reduce_map`. Map Reduce has a different historical meaning that might make this confusing.

Comment (stevenpyzhang) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/map/MapReduce.java]: renamed

**Pull request merged**